### ISSUE TITLE
Dead lock after one epoch, how to sync nodes after estimator.train() return ?

### ISSUE THREAD

**Issue by kzjeef** [horovod/horovod#498]
Hi Guys,

I meet the issue 
```
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
```
I checked my train log, and I found some clue
```
NFO:tensorflow:Loss for final step: 3.3031716.
start wait for other node.
INFO:tensorflow:Loss for final step: 3.2450407.
start wait for other node.
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that di\
fferent ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
Stalled ops:
NmtModel/Distributed_TransformGradients_Allreduce/HorovodAllgather_NmtModel_gradients_NmtModel_Source_Side_WordEmbedding_Gather_grad_Reshape_0 [missing ranks: 3, 6]
NmtModel/Distributed_TransformGradients_Allreduce/HorovodAllreduce_NmtModel_gradients_NmtModel_encoder_layer_0_layer_norm_mul_1_grad_tuple_control_dependency_1_0 [missing ranks: 3, 6]
NmtModel/Distributed_TransformGradients_Allreduce/HorovodAllreduce_NmtModel_gradients_NmtModel_encoder_layer_0_layer_norm_add_1_grad_tuple_control_dependency_1_0 [missing ranks: 3, 6]
NmtModel/Distributed_TransformGradients_Allreduce/HorovodAllreduce_NmtModel_gradients_NmtModel_encoder_layer_0_encoder_self_attention_MatMul_grad_tuple_control_dependency_1_0 [missing ranks: 3\
, 6]
NmtModel/Distributed_TransformGradients_Allreduce/HorovodAllreduce_NmtModel_gradients_NmtModel_encoder_layer_0_encoder_self_attention_add_grad_tuple_control_dependency_1_0 [missing ranks: 3, 6\
]
NmtModel/Distributed_TransformGradients_Allreduc
```

My training code is like the estimator example , but I will shuffle the training input data after each epoch.
so the outer loop is like this:

```
 while True:
        epoch += 1
        if epoch >= params['epoch']//hvd.size():
            break
        tf.logging.info("Epoch %i", epoch)

        if params['shuffle_train']:
            tf.logging.info("Shuffling data for Epoch %i", epoch)
            random.seed(params['seed'])
            states=random.sample(xrange(100000), params['epoch'])
            if not os.path.isfile(train_src+'.epoch'+str(epoch)) or not os.path.isfile(train_trg+'.epoch'+str(epoch)):
                if hvd.local_rank() == 0:
                    shuffle_train(train_src,train_trg,'epoch'+str(epoch),states[epoch])
        train_input_fn = lambda: input_fn(
            train_src+'.epoch'+str(epoch) if params['shuffle_train'] else train_src,
            train_trg+'.epoch'+str(epoch) if params['shuffle_train'] else train_trg,
            vocab_src,
            vocab_trg,
            batch_size_words=params['train_batch_size_words'],
            max_len=params['train_max_len'],
            num_gpus=1,
            is_train=True
        )
        model.train(train_input_fn,  hooks = [bcast_hook])
        print("start wait for other node.")
        time.sleep(100) # add sleep to let multiple process time to finish sync.
```

After one epoch is finished, the train() in the `fast process` is return, but other process is still running the train() function.
Seems like other process thinks this `fast process` was dead, and not able to reduce/broadcast tensor  in the ring.

Seems like it always happens after finish 1 epoch, any idea to fix this issue ?

My MPI command is:
```mpirun -oversubscribe -np 7 -H localhost:7 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH  python train.py
```

**Comment by tgaddair** [comment]
Hey @kzjeef, since this is happening at the end of an epoch, my hunch is that the data is not being evenly divided among the ranks.  Can you check that each rank is receiving the same amount of data in each epoch?
